src/crypto/public_key/rsa.js

```diff
@@ -18,254 +18,205 @@
+      } else if (window.crypto && window.crypto.webkitSubtle) {
         // outdated spec implemented by old Webkit
         keyGenOpt = {
           name: 'RSA-OAEP',
           modulusLength: B, // the specified keysize in bits
-          publicExponent: Euint8.subarray(0, 3), // take three bytes (max 65537)
+          publicExponent: E.toArrayLike(Uint8Array), // take three bytes (max 65537) for exponent
           hash: {
             name: 'SHA-1' // not required for actual RSA keys, but for crypto api 'sign' and 'verify'
           }
         };
-        keys = webCrypto.generateKey(keyGenOpt, true, ['encrypt', 'decrypt']);
-      }
-      else {
+        keyPair = await webCrypto.generateKey(keyGenOpt, true, ['encrypt', 'decrypt']);
+      } else {
         throw new Error('Unknown WebCrypto implementation');
       }
 
-      return keys.then(exportKey).then(function(key) {
-        if (key instanceof ArrayBuffer) {
-          // parse raw ArrayBuffer bytes to jwk/json (WebKit/Safari/IE11 quirk)
-          return decodeKey(JSON.parse(String.fromCharCode.apply(null, new Uint8Array(key))));
-        }
-        return decodeKey(key);
-      });
-    }
-
-    function exportKey(keypair) {
       // export the generated keys as JsonWebKey (JWK)
       // https://tools.ietf.org/html/draft-ietf-jose-json-web-key-33
-      let key = webCrypto.exportKey('jwk', keypair.privateKey);
-      if (typeof key.then !== 'function') { // IE11 KeyOperation
-        key = util.promisifyIE11Op(key, 'Error exporting RSA key pair.');
-      }
-      return key;
-    }
+      let jwk = await webCrypto.exportKey('jwk', keyPair.privateKey);
 
-    function decodeKey(jwk) {
-      // map JWK parameters to local BigInteger type system
-      const key = new KeyObject();
-      key.n = toBigInteger(jwk.n);
-      key.ee = new BigInteger(E, 16);
-      key.d = toBigInteger(jwk.d);
-      key.p = toBigInteger(jwk.p);
-      key.q = toBigInteger(jwk.q);
-      key.u = key.p.modInverse(key.q);
-
-      function toBigInteger(base64url) {
-        const base64 = base64url.replace(/\-/g, '+').replace(/_/g, '/');
-        const hex = util.hexstrdump(atob(base64));
-        return new BigInteger(hex, 16);
+      // parse raw ArrayBuffer bytes to jwk/json (WebKit/Safari/IE11 quirk)
+      if (jwk instanceof ArrayBuffer) {
+        jwk = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(jwk)));
       }
 
+      // map JWK parameters to BN
+      key = {};
+      key.n = b64toBN(jwk.n);
+      key.e = E;
+      key.d = b64toBN(jwk.d);
+      key.p = b64toBN(jwk.p);
+      key.q = b64toBN(jwk.q);
+      key.u = key.p.invm(key.q);
       return key;
     }
 
-    //
-    // JS code
-    //
-
-    return new Promise(function(resolve) {
-      const key = new KeyObject();
-      const rng = new SecureRandom();
-      const qs = B >> 1;
-      key.e = parseInt(E, 16);
-      key.ee = new BigInteger(E, 16);
+    while (true) {
```

Comment:
Why is this in a while loop?

Reply:
Ah, good catch. The previous implementation would restart if certain checks failed, but those checks are all in `randomProbablePrime` now.

@sanjanarajan do you know if there is any case where two primes would not be good to use together? Something about common factors of `p-1` and `q-1` perhaps?